utils/global-state-update-gen/src/generic.rs

```diff
@@ -0,0 +1,284 @@
+/// snaphot according to the supplied list of configured accounts.
+fn gen_snapshot_from_old(
+    mut snapshot: SeigniorageRecipientsSnapshot,
+    accounts: &[AccountConfig],
+) -> SeigniorageRecipientsSnapshot {
+    let stakes_map: BTreeMap<_, _> = accounts
+        .iter()
+        .filter_map(|acc| acc.stake.map(|stake| (acc.public_key.clone(), stake)))
+        .collect();
+
+    for recipients in snapshot.values_mut() {
+        let mut to_drop = vec![];
+        for (public_key, recipient) in recipients.iter_mut() {
+            match stakes_map.get(public_key) {
+                Some(stake) if *stake == U512::zero() => {
+                    to_drop.push(public_key.clone());
+                }
+                Some(stake) => {
+                    *recipient =
+                        SeigniorageRecipient::new(*stake, Default::default(), Default::default());
+                }
+                None => (),
+            }
+        }
+        for public_key in to_drop {
+            recipients.remove(&public_key);
+        }
+    }
+
+    snapshot
+}
+
+/// Generates a set of writes necessary to "fix" the bids, ie.:
+/// - set the bids of the new validators to their desired stakes,
+/// - remove the bids of the old validators that are no longer validators,
+/// - remove all the bids that are larger than the smallest bid among the new validators
+/// (necessary, because such bidders would outbid the validators decided by the social
+/// consensus).
+pub fn add_and_remove_bids<T: StateReader>(
+    state: &mut StateTracker<T>,
+    validators_diff: &ValidatorsDiff,
+    new_snapshot: &SeigniorageRecipientsSnapshot,
+    only_listed_validators: bool,
+) {
+    let to_unbid = if only_listed_validators {
+        let large_bids = find_large_bids(state, new_snapshot);
+        validators_diff
+            .removed
+            .union(&large_bids)
+            .cloned()
+            .collect()
+    } else {
+        validators_diff.removed.clone()
+    };
+
+    for (pub_key, seigniorage_recipient) in new_snapshot.values().next().unwrap() {
+        let stake = *seigniorage_recipient.stake();
+        create_or_update_bid(state, pub_key, stake);
+    }
+
+    // Refresh the bids - we modified them above.
+    let bids = state.get_bids();
+
+    for pub_key in to_unbid {
+        if let Some(bid) = bids.get(&pub_key) {
+            let new_bid = Bid::empty(pub_key.clone(), *bid.bonding_purse());
+            state.set_bid(pub_key.clone(), new_bid);
+        }
+    }
+}
+
+/// Returns the set of public keys that have bids larger than the smallest bid among the new
+/// validators.
+fn find_large_bids<T: StateReader>(
+    state: &mut StateTracker<T>,
+    snapshot: &SeigniorageRecipientsSnapshot,
+) -> BTreeSet<PublicKey> {
+    let seigniorage_recipients = snapshot.values().next().unwrap();
+    let min_bid = seigniorage_recipients
+        .values()
+        .map(SeigniorageRecipient::stake)
```

Comment:
Maybe in case we later change this code to add or keep delegators?

```suggestion
        .map(SeigniorageRecipient::total_stake)
```

Reply:
Fixed in [8c25879](https://github.com/casper-network/casper-node/pull/3195/commits/8c25879b795846909ecad82d3f0d2b28ed137eb8)